### ISSUE TITLE
Agent crash: fatal error: concurrent map read and map write

### ISSUE THREAD

**Issue by DMarby** [grafana/agent#3910]
### What's wrong?

Agent crashes periodically with a stacktrace

### Steps to reproduce

Run grafana-agent with the given configuration.
Agent is deployed as a replicaset in Kubernetes using the helm chart.

### Software version

Helm version 0.12/Grafana Agent 0.33.0

### Configuration

```text
logging {
  level  = "info"
  format = "json"
}


loki.source.gcplog "logs" {
  pull {
    project_id   = "<project id>"
    subscription = "<subscription>"
    use_incoming_timestamp = true
    labels = {
      job = "gcp",
    }
  }

  forward_to = [
    loki.process.processor_one.receiver,
    loki.process.processor_two.receiver,
    loki.write.loki_write.receiver,
  ]
}


loki.process "processor_one" {
  forward_to = []

  stage.match {
    selector = "{job=\"gcp\"} |~ \".*type.googleapis.com/google.cloud.loadbalancing.type.LoadBalancerLogEntry.*\" |~ \".*(?P<scheme>https:|http:)//(?P<domain>[a-zA-Z0-9_.]+)/(?P<custom_label_one>.[.a-zA-Z0-9_]+)/(?P<custom_label_two>[A-Z][a-zA-Z0-9_]+).*\""

    stage.json {
        expressions = {
          status  = "httpRequest.status",
          latency = "httpRequest.latency",
        }
    }

    stage.regex {
      expression = "(?P<scheme>https:|http:)//(?P<domain>[a-zA-Z0-9_.]+)/(?P<custom_label_one>.[.a-zA-Z0-9_]+)/(?P<custom_label_two>[A-Z][a-zA-Z0-9_]+)\""
    }

    stage.labels {
      values = {
        custom_label_one = "",
        custom_label_two  = "",
        status      = "",
      }
    }

    stage.metrics {
      metric.counter {
        name        = "count"
        prefix      = "metric_one_"
        match_all   = true

        action            = "inc"
        max_idle_duration = "5m"
      }

      metric.histogram {
        name        = "latency_seconds"
        prefix      = "metric_one_"

        source = "latency"
        buckets = [
          0.005, // 5 ms
          0.01,  // 10 ms
          0.025, // 25 ms
          0.050, // 50 ms
          0.075, // 75 ms
          0.1,   // 100 ms
          0.25,  // 250 ms
          0.5,   // 500 ms
          0.75,  // 750 ms
          1.0,   // 1 s
          2.5,   // 2.5 s
          5.0,   // 5 s
          7.5,   // 7.5 s
          10.0,  // 10 s
          25.0,  // 25 s
          50.0,  // 50 s
          75.0,  // 1m15s
          100.0, // 1m40s
          250.0, // 4m10s
        ]

        max_idle_duration = "5m"
      }
    }
  }
}


loki.process "processor_two" {
  forward_to = []

  stage.match {
    selector = "{job=\"gcp\"} |~ \".*type.googleapis.com/google.cloud.loadbalancing.type.LoadBalancerLogEntry.*\" |~ \".*requestUrl.*?(?P<scheme>https:|http:)//(?P<domain>[a-zA-Z0-9_.]+)/(?P<custom_label>[a-zA-Z0-9]+).*\""

    stage.json {
        expressions = {
          status  = "httpRequest.status",
          latency = "httpRequest.latency",
        }
    }

    stage.regex {
      expression = "requestUrl.*?(?P<scheme>https:|http:)\\/\\/(?P<domain>[a-zA-Z0-9_.]+)\\/(?P<custom_label>[a-zA-Z0-9]+)"
    }

    stage.labels {
      values = {
        status        = "",
        custom_label = "",
      }
    }

    stage.metrics {
      metric.histogram {
        name        = "latency_seconds"
        prefix      = "metric_two_"

        source = "latency"
        buckets = [
          0.005, // 5 ms
          0.01,  // 10 ms
          0.025, // 25 ms
          0.050, // 50 ms
          0.075, // 75 ms
          0.1,   // 100 ms
          0.25,  // 250 ms
          0.5,   // 500 ms
          0.75,  // 750 ms
          1.0,   // 1 s
          2.5,   // 2.5 s
          5.0,   // 5 s
          7.5,   // 7.5 s
          10.0,  // 10 s
          25.0,  // 25 s
          50.0,  // 50 s
          75.0,  // 1m15s
          100.0, // 1m40s
          250.0, // 4m10s
        ]

        max_idle_duration = "5m"
      }
    }
  }
}


loki.write "loki_write" {
  external_labels = {
    cluster = "<cluster>",
  }

  endpoint {
    url = "<loki endpoint>"

    basic_auth {
      username = "<username>"
      password_file = "/config/secrets/secret"
    }
  }
}

prometheus.scrape "self" {
  job_name = "grafana-agent"
  scrape_interval = "30s"

  targets = [
    {"__address__" = "127.0.0.1:80", "instance" = constants.hostname},
  ]

  forward_to = [prometheus.relabel.keep_log_metrics_only.receiver]
}

prometheus.relabel "keep_log_metrics_only" {
  rule {
    action        = "keep"
    source_labels = ["__name__"]
    regex         = "loki_source_gcplog_.*|loki_write_.*|load_balancer_.*"
  }

  forward_to = [prometheus.remote_write.remote_write.receiver]
}

prometheus.remote_write "remote_write" {
  external_labels = {
    cluster = "<cluster>",
  }

  endpoint {
    url = "https://<prom api>/api/prom/push"

    basic_auth {
      username = "<username>"
      password_file = "/config/secrets/secret"
    }
  }
}
```

### Logs

```text
fatal error: concurrent map read and map write
1460
1459
goroutine 259 [running]:
1458
github.com/grafana/agent/component/loki/process/internal/stages.(*matcherStage).processLogQL(0xc000c6f620, {0xc0034a1770, {0xc003332d50, {{0x8b3a558, 0xedbf69d0d, 0x0}, {0xc002fd5180, 0x57e}}}})
1457
	/src/agent/component/loki/process/internal/stages/match.go:188 +0xd6
1456
github.com/grafana/agent/component/loki/process/internal/stages.(*matcherStage).runKeep.func2()
1455
	/src/agent/component/loki/process/internal/stages/match.go:160 +0x173
1454
created by github.com/grafana/agent/component/loki/process/internal/stages.(*matcherStage).runKeep
1453
	/src/agent/component/loki/process/internal/stages/match.go:157 +0x145
1452
1451
goroutine 1 [select, 11 minutes]:
1450
github.com/grafana/agent/cmd/internal/flowmode.(*flowRun).Run(0xc000f4fd40, {0x7fffc179a594, 0x17})
1449
	/src/agent/cmd/internal/flowmode/cmd_run.go:302 +0x1425
1448
github.com/grafana/agent/cmd/internal/flowmode.runCommand.func1(0xc0002def00?, {0xc000d7c940?, 0x4?, 0x4?})
1447
	/src/agent/cmd/internal/flowmode/cmd_run.go:79 +0x34
1446
github.com/spf13/cobra.(*Command).execute(0xc0002def00, {0xc000d7c800, 0x4, 0x4})
1445
	/go/pkg/mod/github.com/spf13/cobra@v1.6.1/command.go:916 +0x862
1444
github.com/spf13/cobra.(*Command).ExecuteC(0xc0002de900)
1443
	/go/pkg/mod/github.com/spf13/cobra@v1.6.1/command.go:1044 +0x3bd
1442
github.com/spf13/cobra.(*Command).Execute(...)
1441
	/go/pkg/mod/github.com/spf13/cobra@v1.6.1/command.go:968
1440
github.com/grafana/agent/cmd/internal/flowmode.Run()
1439
	/src/agent/cmd/internal/flowmode/flowmode.go:31 +0x19f
1438
main.main()
1437
	/src/agent/cmd/grafana-agent/main.go:51 +0x605
1436
1435
goroutine 65 [select]:
1434
go.opencensus.io/stats/view.(*worker).start(0xc00029f180)
1433
	/go/pkg/mod/go.opencensus.io@v0.24.0/stats/view/worker.go:292 +0xad
1432
created by go.opencensus.io/stats/view.init.0
1431
	/go/pkg/mod/go.opencensus.io@v0.24.0/stats/view/worker.go:34 +0x96
1430
1429
goroutine 214 [chan receive]:
1428
github.com/ClickHouse/clickhouse-go.init.0.func1()
1427
	/go/pkg/mod/github.com/!click!house/clickhouse-go@v1.5.4/bootstrap.go:48 +0x2d
1426
created by github.com/ClickHouse/clickhouse-go.init.0
1425
	/go/pkg/mod/github.com/!click!house/clickhouse-go@v1.5.4/bootstrap.go:45 +0x3f
1424
1423
goroutine 183 [select, 11 minutes]:
1422
github.com/grafana/agent/cmd/internal/flowmode.interruptContext.func1()
1421
	/src/agent/cmd/internal/flowmode/cmd_run.go:345 +0xf4
1420
created by github.com/grafana/agent/cmd/internal/flowmode.interruptContext
1419
	/src/agent/cmd/internal/flowmode/cmd_run.go:341 +0xa5
1418
1417
goroutine 184 [select]:
1416
go.opentelemetry.io/otel/sdk/trace.(*batchSpanProcessor).processQueue(0xc001586320)
1415
	/go/pkg/mod/go.opentelemetry.io/otel/sdk@v1.11.2/trace/batch_span_processor.go:293 +0x12c
1414
go.opentelemetry.io/otel/sdk/trace.NewBatchSpanProcessor.func1()
1413
	/go/pkg/mod/go.opentelemetry.io/otel/sdk@v1.11.2/trace/batch_span_processor.go:128 +0x65
1412
created by go.opentelemetry.io/otel/sdk/trace.NewBatchSpanProcessor
1411
	/go/pkg/mod/go.opentelemetry.io/otel/sdk@v1.11.2/trace/batch_span_processor.go:126 +0x2ea
1410
1409
goroutine 185 [chan receive, 11 minutes]:
1408
github.com/grafana/agent/pkg/flow/tracing.(*Tracer).Run(0xc000b7c480, {0x79a94b0, 0xc000597360})
1407
	/src/agent/pkg/flow/tracing/tracing.go:194 +0x6a
1406
github.com/grafana/agent/cmd/internal/flowmode.(*flowRun).Run.func1()
1405
	/src/agent/cmd/internal/flowmode/cmd_run.go:137 +0x32
1404
created by github.com/grafana/agent/cmd/internal/flowmode.(*flowRun).Run
1403
	/src/agent/cmd/internal/flowmode/cmd_run.go:136 +0x31f
1402
```

**Comment by tpaschalis** [comment]
Thanks for the report. We recently had a similar issue in #3823 and the loki.source.kubernetes component; I'm taking a look.

**Comment by tpaschalis** [comment]
@DMarby I _think_ the issue has a similar solution, but I'd like to verify it.

Is what you posted the entire stack trace after `fatal error: concurrent map read and map write`, or could there be more lines in it?

**Comment by DMarby** [comment]
> Is what you posted the entire stack trace after `fatal error: concurrent map read and map write`, or could there be more lines in it?

Sorry about that, it seems like it got cut off, this should be the full stacktrace: https://gist.github.com/DMarby/2a2612e0f404a66d5e7fb51887f8e632

**Comment by tpaschalis** [comment]
Thanks for the complete stack trace.

Unfortunately I think the issue here is more complex than #3823 and applicable to more components than one.

When we're pushing to multiple receivers, we're fanning out the same log entry to be handled by stages in different goroutines, leading to the concurrent reads and writes on the label map.

This would be fixed for all components by pushing on for #2929 as a shared way to handle log entries. As a temporary workaround, do you think the two loki.process stages could be made into one by utilizing a more complex selector and regex, and merging the metric stages?

**Comment by DMarby** [comment]
> do you think the two loki.process stages could be made into one by utilizing a more complex selector and regex, and merging the metric stages?

This seems to have stopped the crashing, thank you!
